File: neural java/NeuralNetwork.java
package itu.ejuu.neural;

import java.util.ArrayList;
import java.util.Random;

public class NeuralNetwork {

	public static int RESTARTS = 0;
	public static int LIMIT = 10000; //Integer.MAX_VALUE;
	public static boolean DEBUG = false;
	
	public static void main(String[] args){
		
		// create training pairs XOR
		double[][] trainingInputs = new double[][]{new double[]{0,0},new double[]{1,0},new double[]{0,1},new double[]{1,1}};
		double[][] trainingOutputs = new double[][]{new double[]{0},new double[]{1},new double[]{1},new double[]{0}};
		
		int restarts = 0;
		
		while(restarts <= RESTARTS){ // input,output,hidden,levels,outputValues
			NeuralNetwork nn = new NeuralNetwork(2,1,2,1);

			int i = 0;
			boolean allCorrect = false;
			
			while(!allCorrect && i < LIMIT){ // stopping condition
				System.out.println("----- Running Training "+i+" -----");
				allCorrect = nn.runTraining( trainingInputs
											,trainingOutputs) == trainingInputs.length;
				i++;
			}
			
			if(allCorrect){
				System.out.println("Complete after "+i+" runs.");
				System.out.println(nn.toString());
				break;
			}else{
				restarts++;
				if(restarts <= RESTARTS){
					System.out.println("Restart #"+restarts);
				}else{
					System.out.println("Max restarts exceeded.");
				}
				
			}
		}
	}
	
	// -------------------------------------
	
	private ArrayList<Node> inputNodes, outputNodes;
	private ArrayList<ArrayList<Node>> hiddenLayers;
	private ArrayList<Connection> connections;
	
	public NeuralNetwork(int numInput, int numOutput, int numHidden, int numLayers){
		Random rand = new Random();
		
		// create tree
		inputNodes = new ArrayList<Node>();
		hiddenLayers = new ArrayList<ArrayList<Node>>(numLayers);
		outputNodes = new ArrayList<Node>();
		connections = new ArrayList<Connection>();
		
		for(int i = 0; i < numInput; i++) inputNodes.add(new Node(randStart(rand)));
		for(int i = 0; i < numOutput; i++) outputNodes.add(new Node(randStart(rand)));
		for(int i = 0; i < numLayers; i++){
			hiddenLayers.add(new ArrayList<Node>(numHidden));
			for(int j = 0; j < numHidden; j++){
				Node curNode = new Node(randStart(rand));
				hiddenLayers.get(i).add(curNode);
				// hook up to previous
				if(i == 0){ // first layer to input
					for(int k = 0; k < numInput; k++){
						connections.add(new Connection(randStart(rand), inputNodes.get(k), curNode));
					}
				}else{ // rest to layer before
					for(int k = 0; k < numHidden; k++){
						connections.add(new Connection(randStart(rand), hiddenLayers.get(i-1).get(k), curNode));
					}
				}
				// to output
				if(i == numLayers-1){
					for(int k = 0; k < numOutput; k++){
						connections.add(new Connection(randStart(rand), curNode, outputNodes.get(k)));
					}
				}
			}
		}
	}
	
	private double randStart(Random rand){
		return (rand.nextDouble()*2.0)-1.0;
	}
	
	public double[] getOutput(double[] parseInput) {
		double[] result = new double[outputNodes.size()];
		for(int i = 0; i < result.length; i++){
			result[i] = outputNodes.get(i).forwardOperation();
		}
		return result;
	}
	
	public boolean runTraining(double[] trainingInput, double[] trainingOutput){
		// set inputs
		for(int j = 0; j < inputNodes.size(); j++){
			inputNodes.get(j).setInput(trainingInput[j]);
		}
		
		if(DEBUG) System.out.println("Output=");
		boolean match = true;
		for(int j = 0; j < outputNodes.size(); j++){
			// forward operation
			double result = outputNodes.get(j).forwardOperation();
			int round = (int)Math.round(result);
			int target = (int)Math.round(trainingOutput[j]);
			
			if(DEBUG) System.out.println("\t"+j+": "+result+" - Target= "+trainingOutput[j]);
			if(DEBUG) System.out.println("\t"+j+"(int): "+round+" - Target(int)= "+target);
			
			match = match && (target == round);
		}
	
		if(DEBUG) System.out.println("Result Match: " + match);
		
		// output error
		for(int j = 0; j < outputNodes.size(); j++){
			double T = trainingOutput[j];
			outputNodes.get(j).computeError(T);
		}
		for(int j = hiddenLayers.size()-1; j >= 0; j--){
			for(Node n : hiddenLayers.get(j)) n.computeError(0);
		}
		
		// update weights
		for(Connection c : connections) c.updateWeight();
		
		// update bias
		for(int j = 0; j < hiddenLayers.size(); j++){
			for(Node n : hiddenLayers.get(j)) n.updateBias();
		}
		for(Node n : outputNodes) n.updateBias();
		
		if(DEBUG) System.out.println(this);
		
		return match;
	}

	public int runTraining(double[][] trainingInput, double[][] trainingOutput) {
		int numCorrect = 0;
		for(int i = 0; i < trainingInput.length; i++){
			boolean match = runTraining(trainingInput[i], trainingOutput[i]);
			if(match) numCorrect++;
		}
		if(DEBUG) System.out.println(numCorrect+" out of "+trainingInput.length+" correct.");
		return numCorrect;
	}
	
	private void nodeToBuilder(Node n, StringBuilder builder){
		builder.append(n+": bias="+n.getBias()+"\n");
		for(Connection c : n.conn_out){
			builder.append("\t-> "+c.getToNode()+" = "+c.getWeight()+"\n");
		}
	}
	
	@Override
	public String toString(){
		StringBuilder builder = new StringBuilder();
		builder.append("Input Nodes:\n");
		for(Node n : inputNodes) nodeToBuilder(n,builder);
		
		for(int i = 0; i < hiddenLayers.size(); i++){
			builder.append("\nLayer "+(i+1)+":\n");
			for(Node n : hiddenLayers.get(i)) nodeToBuilder(n,builder);
		}

		builder.append("\nOutput Nodes:\n");
		for(Node n : outputNodes) nodeToBuilder(n,builder);
		builder.append("\n");
		return builder.toString();
	}
}

File: DataMining/Neural Networks/NeuralNetwork.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace DataMining.Neural_Networks
{
    public class NeuralNetwork
    {
        public const bool DEBUG = false;

        private Node[] inputNodes, outputNodes;
        private Node[][] hiddenLayers;
        private List<Connection> connections;

        private readonly Random rand;

        public NeuralNetwork(int inputNum, int outputNum, int hiddenNum, int layerNum)
        {
            rand = new Random();

            connections = new List<Connection>();
            hiddenLayers = new Node[layerNum][];
            outputNodes = new Node[outputNum];
            inputNodes = new Node[inputNum];

            for (int i = 0; i < inputNum; i++)
            {
                inputNodes[i] = new Node(RandStart());
            }

            for (int i = 0; i < outputNum; i++)
            {
                outputNodes[i] = new Node(RandStart());
            }
            for (int i = 0; i < layerNum; i++)
            {
                hiddenLayers[i] = new Node[hiddenNum];
                for (int j = 0; j < hiddenNum; j++)
                {
                    Node curNode = new Node(RandStart());
                    hiddenLayers[i][j] = curNode;

                    //hook up connection
                    if (i == 0) //input layer to first hiddden layer
                    {
                        for (int k = 0; k < inputNum; k++)
                        {
                            connections.Add(new Connection(RandStart(), inputNodes[k], curNode));

                        }
                    }
                    else //hidden layer to hidden layer
                    {
                        for (int k = 0; k < inputNum; k++)
                        {
                            connections.Add(new Connection(RandStart(), hiddenLayers[i - 1][k], curNode));

                        }
                    }

                    //last hidden layer to output layer
                    if (i == layerNum - 1)
                    {
                        for (int k = 0; k < outputNum; k++)
                        {
                            connections.Add(new Connection(RandStart(), curNode, outputNodes[k]));
                        }
                    }
                }
            }
        }

        public double RandStart()
        {
            return (rand.NextDouble()*2)-1.0;
        }

        public double[] CalculateOutput(double[] inputVariables)
        {
            double[] result = new double[outputNodes.Length];
            for (int i = 0; i < result.Length; i++)
            {
                result[i] = outputNodes[i].ForwardOperation();
            }
            return result;
        }

        public bool RunTraining(double[] trainingInput, double[] trainingOutput)
        {
            //set inputs
            for (int i = 0; i < inputNodes.Length; i++)
            {
                inputNodes[i].InputValue = trainingInput[i];
            }

            if(DEBUG) Console.WriteLine("Output=");
            bool match = true;
            for (int i = 0; i < outputNodes.Length; i++)
            {
                //forward operation
                double result = outputNodes[i].ForwardOperation();
                int rounded = (int) Math.Round(result);
                int target = (int) Math.Round(trainingOutput[i]);

                if (DEBUG) Console.WriteLine("\t{0}: {1} - Target={2}", i, result, trainingOutput[i]);
                if (DEBUG) Console.WriteLine("\t{0}(int): {1} - Target(int)={2}", i, rounded, target);

                match = match && (target == rounded);
            }

            if (DEBUG) Console.WriteLine("Result Match: {0}", match);

            // output error
            for (int i = 0; i < outputNodes.Length; i++)
            {
                double T = trainingOutput[i];
                outputNodes[i].ComputeError(T);
            }
            for (int i = hiddenLayers.Length-1; i >= 0; i--)
            {
                foreach (Node hiddenNode in hiddenLayers[i])
                {
                    hiddenNode.ComputeError(0);
                }
            }

            //update weights
            connections.ToList().ForEach(e => e.UpdateWeight());

            //update bias
            foreach (Node hNode in hiddenLayers.SelectMany(hLayer => hLayer))
            {
                hNode.UpdateBias();
            }
            outputNodes.ToList().ForEach(e => e.UpdateBias());

            if(DEBUG) Console.WriteLine(this);

            return match;
        }

        public int runSession(double[][] trainingInput, double[][] trainingOutput)
        {
            int numCorrect = 0;

            for (int i = 0; i < trainingInput.Length; i++)
            {
                bool match = RunTraining(trainingInput[i], trainingOutput[i]);
                if (match) numCorrect++;
                
            }
            if(DEBUG) Console.WriteLine("{0} out of {1} correct", numCorrect, trainingInput.Length);
            return numCorrect;
        }

        public override string ToString()
        {
            StringBuilder strB = new StringBuilder();
            strB.Append("Input Nodes:\n");
            inputNodes.ToList().ForEach(e => NodeToBuilder(e, strB));
        
            for (int i = 0; i < hiddenLayers.Length; i++)
            {
                strB.AppendFormat("\nLayer {0}:\n", i + 1);
                hiddenLayers[i].ToList().ForEach(e => NodeToBuilder(e,strB));
            }

            strB.Append("\nOutput Nodes\n");
            outputNodes.ToList().ForEach(e => NodeToBuilder(e,strB));

            strB.Append("\n");
            return strB.ToString();
        }

        private void NodeToBuilder(Node node, StringBuilder strB)
        {
            strB.AppendFormat("{0}: bias={1}\n",node, node.Bias);
            foreach (Connection conn in node.OutCon)
            {
                strB.AppendFormat("\t-> {0} = {1}\n", conn.To, conn.Weight);
            }
        }
    }
}
